File: Testing/tests/client/Tests/Linq/TestLinqQueryExecution.cs
using Bridge.Linq;
using Bridge.QUnit;
using System.Linq;

namespace ClientTestLibrary.Linq
{
    internal class TestLinqQueryExecution
    {
        public static void Test(Assert assert)
        {
            assert.Expect(6);

            // TEST
            int[] numbers = new int[] { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
            int i = 0;

            var aQuery = from n in numbers
                         select ++i;
            assert.Equal(i, 0, "Query is not executed until you enumerate over them");

            // TEST
            aQuery.ToList();
            assert.Equal(i, 10, "Query is  executed after you enumerate over them");

            i = 0;

            // TEST
            var bQuery = (from n in numbers
                          select ++i).Max();
            assert.Equal(i, 10, "Max() executes immediately");

            // TEST
            var smallNumbers = from n in numbers
                               where n <= 3
                               select n;
            var smallerEvenNumbers = from n in smallNumbers
                                     where n % 2 == 0
                                     select n;
            assert.DeepEqual(smallerEvenNumbers.ToArray(), new[] { 2, 0 }, "Query in a query");

            // TEST
            numbers.ForEach((x, index) => numbers[index] = -numbers[index]);
            assert.DeepEqual(numbers.ToArray(), new int[] { -5, -4, -1, -3, -9, -8, -6, -7, -2, 0 }, "ForEach()");

            // TEST
            assert.DeepEqual(smallerEvenNumbers.ToArray(), new[] { -4, -8, -6, -2, 0 }, "Second query run on a modified source");
        }
    }
}

File: Testing/www/js/testLinqQueryExecution.js
/* global Bridge */

"use strict";

Bridge.define('ClientTestLibrary.Linq.TestLinqQueryExecution', {
    statics: {
        test: function (assert) {
            assert.expect(6);

            // TEST
            var numbers = [5, 4, 1, 3, 9, 8, 6, 7, 2, 0];
            var i = 0;

            var aQuery = Bridge.Linq.Enumerable.from(numbers).select(function (n) {
                return ++i;
            });
            assert.equal(i, 0, "Query is not executed until you enumerate over them");

            // TEST
            aQuery.toList(Bridge.Int);
            assert.equal(i, 10, "Query is  executed after you enumerate over them");

            i = 0;

            // TEST
            var bQuery = (Bridge.Linq.Enumerable.from(numbers).select(function (n) {
                return ++i;
            })).max();
            assert.equal(i, 10, "Max() executes immediately");

            // TEST
            var smallNumbers = Bridge.Linq.Enumerable.from(numbers).where(function (n) {
                return n <= 3;
            });
            var smallerEvenNumbers = smallNumbers.where(function (n) {
                return n % 2 === 0;
            });
            assert.deepEqual(smallerEvenNumbers.toArray(), [2, 0], "Query in a query");

            // TEST
            Bridge.Linq.Enumerable.from(numbers).forEach(function (x, index) {
                return numbers[index] = -numbers[index];
            });
            assert.deepEqual(Bridge.Linq.Enumerable.from(numbers).toArray(), [-5, -4, -1, -3, -9, -8, -6, -7, -2, 0], "ForEach()");

            // TEST
            assert.deepEqual(smallerEvenNumbers.toArray(), [-4, -8, -6, -2, 0], "Second query run on a modified source");
        }
    }
});



Bridge.init();